draft/XXXX-multipart-upload/README.md

```diff
@@ -0,0 +1,115 @@
+---
+RFC: XXXX
+Author: Roni Choudhury
+Status: draft
+Created: 2020-08-04
+Last-Modified: 2020-08-04
+---
+
+# RFC XXXX: Multipart Table Upload
+
+This RFC describes how large table data can be safely and efficiently uploaded
+to Multinet, using a variation of the common multipart upload strategy common to
+many web server technologies.
+
+## Motivation
+
+Multipart upload is needed for the following reasons:
+
+1. As a practical matter, our hosting platform Heroku does not allow for
+   requests that take longer than thirty seconds to complete. Some moderately
+   sized CSV files already break this threshold, and so called "big data" tables
+   most certainly will. In order to upload such data files, we will need a way
+   to upload them in chunks.
+2. A long-running one-shot upload does not give the end user much feedback on
+   what is going on. Multipart upload would allow us to provide a more detailed
+   progress meter that can improve the user experience.
+3. If a long-running upload is interrupted for any reason, a multipart upload
+   strategy allows for resuming such uploads without losing the time already
+   spent in uploading one more initial chunks of the file.
+
+## Proposals
+
+### Summary of Approach
+
+The basic idea is as follows: the client initiates a new upload session for a
+given file; this action causes the server to create a new collection in the
+database. Then the client issues one or more "chunk" uploads, each sending a
+range of bytes, up to some size limit, to the server. The server saves each
+chunk as a document in the database collection associated with this file upload.
+Finally, the client issues a "finalize" call, which marks the database
+collection as being complete.
+
+The server can then ask for a file-like object associated with the upload. The
+server can then use this object to perform validation, retrieve table rows, and
+settle them in their final home in the server (as is done currently, directly
+with the file contents).
+
+Once the work is completed, the server can delete the upload collection from the
+database.
+
+### API design
+
+This RFC adds an `uploads` section to the Multinet API, which contains the
+following endpoints:
+
+- `POST /api/uploads` - creates a new upload object by selecting a random,
+non-preexisting SHA string, then creating an Arango collection keyed by the
+string. This endpoint returns the ID string.
+- `POST /api/uploads/<upload-id>/chunk?sequence=<sequence-number>` - uploads one
+chunk of data to the named upload object. This endpoint returns the sequence
+number, and may raise various status codes to indicate problems.
+- `POST /api/uploads/<upload-id>/finalize` - finalizes the IDed upload by
+preventing further calls to `POST /api/uploads/<upload-id>/chunk` and performing
+a checksum on the settled data.
```

Comment:
What happens if a chunk upload fails half way? Would there be a hanging part? How would the API stitch together the file with these half uploaded chunks?

Also, what is the checksum here? A checksum for each chunk? Maybe that should be ran on each post of a chunk

Reply:
Chunks are all-or-nothing, so if a chunk upload fails halfway, then nothing is written to the database, and the client receives an error status code.

There are block-based checksums that can be run by streaming the target data. My thinking was to compute such a checksum and compare it to one sent at some point during this whole process.

Your idea for running checksums on each chunk is not a bad one. However, to make use of it, the client would need to compute a checksum for each chunk, and send that value along with the chunk upload. That's a very unnatural way to operate. Usually, the client is interested in verifying the checksum for the full data. That can certainly be computed once and sent along with the `finalize` call.

Reply:
I think checking the whole file is okay, I just wonder if it will be easy to compute for the whole file if it's broken up into chunks

Reply:
> I think checking the whole file is okay, I just wonder if it will be easy to compute for the whole file if it's broken up into chunks

For block-based checksums it should be easy (though I don't have experience with this). If there's a Python library that does it, I imagine simply passing an iterator to a function would "just work". This is something we need to do a bit of research into.